File: app/src/main/java/com/github/shadowsocks/utils/Parser.java
package com.github.shadowsocks.utils;

import android.text.TextUtils;
import android.util.Base64;

import com.github.shadowsocks.ShadowsocksApplication;
import com.github.shadowsocks.database.Profile;

import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Parser {

    public static final String TAG = Parser.class.getSimpleName();

    private static final String pattern_regex = "(?i)ss://([A-Za-z0-9+-/=_]+)(#(.+))?";
    private static final String decodedPattern_regex = "(?i)^((.+?)(-auth)??:(.*)@(.+?):(\\d+?))$";

    private static final String pattern_ssr_regex = "(?i)ssr://([A-Za-z0-9_=-]+)";
    private static final String decodedPattern_ssr_regex = "(?i)^((.+):(\\d+?):(.*):(.+):(.*):([^/]+))";
    private static final String decodedPattern_ssr_obfsparam_regex = "(?i)[?&]obfsparam=([A-Za-z0-9+-/=_]*)";
    private static final String decodedPattern_ssr_remarks_regex = "(?i)[?&]remarks=([A-Za-z0-9+-/=_]*)";
    private static final String decodedPattern_ssr_protocolparam_regex = "(?i)[?&]protoparam=([A-Za-z0-9+-/=_]*)";
    private static final String decodedPattern_ssr_groupparam_regex = "(?i)[?&]group=([A-Za-z0-9+-/=_]*)";

    private static Pattern getPattern(String regex) {
        return Pattern.compile(regex);
    }

    public static List<Profile> findAll(CharSequence data) {
        Pattern pattern = getPattern(pattern_regex);
        Pattern decodedPattern = getPattern(decodedPattern_regex);

        CharSequence input = null;
        if (!TextUtils.isEmpty(data)) {
            input = data;
        } else {
            input = "";
        }
        Matcher m = pattern.matcher(input);
        try {
            List<Profile> list = new ArrayList<>();
            while (m.find()) {
                Matcher ss = decodedPattern.matcher(new String(Base64.decode(m.group(1), Base64.NO_PADDING), StandardCharsets.UTF_8));
                if (ss.find()) {
                    Profile profile = new Profile();
                    profile.setMethod(ss.group(2).toLowerCase());
                    if (ss.group(3) != null) {
                        profile.setProtocol("verify_sha1");
                    }
                    profile.setPassword(ss.group(4));
                    profile.setName(ss.group(5));
                    profile.setHost(profile.getName());
                    profile.setRemotePort(Integer.parseInt(ss.group(6)));
                    if (m.group(2) != null) {
                        profile.setName(URLDecoder.decode(m.group(3), "utf-8"));
                    }
                    list.add(profile);
                }
            }
            return list;
        } catch (Exception e) {
            // Ignore
            VayLog.INSTANCE.e(TAG, "findAll", e);
            ShadowsocksApplication.app.track(e);
            return null;
        }
    }

    public static List<Profile> findAll_ssr(CharSequence data) {
        Pattern pattern_ssr = getPattern(pattern_ssr_regex);
        Pattern decodedPattern_ssr = getPattern(decodedPattern_ssr_regex);
        Pattern decodedPattern_ssr_obfsparam = getPattern(decodedPattern_ssr_obfsparam_regex);
        Pattern decodedPattern_ssr_protocolparam = getPattern(decodedPattern_ssr_protocolparam_regex);
        Pattern decodedPattern_ssr_remarks = getPattern(decodedPattern_ssr_remarks_regex);
        Pattern decodedPattern_ssr_groupparam = getPattern(decodedPattern_ssr_groupparam_regex);

        CharSequence input = null;
        if (!TextUtils.isEmpty(data)) {
            input = data;
        } else {
            input = "";
        }

        Matcher m = pattern_ssr.matcher(input);
        try {
            List<Profile> list = new ArrayList<>();
            while (m.find()) {
                String uri = new String(Base64.decode(m.group(1).replaceAll("=", ""), Base64.URL_SAFE), StandardCharsets.UTF_8);
                Matcher ss = decodedPattern_ssr.matcher(uri);
                if (ss.find()) {
                    Profile profile = new Profile();
                    profile.setHost(ss.group(2).toLowerCase());
                    profile.setRemotePort(Integer.parseInt(ss.group(3)));
                    profile.setProtocol(ss.group(4).toLowerCase());
                    profile.setMethod(ss.group(5).toLowerCase());
                    profile.setObfs(ss.group(6).toLowerCase());
                    profile.setPassword(new String(Base64.decode(ss.group(7).replaceAll("=", ""), Base64.URL_SAFE), StandardCharsets.UTF_8));

                    Matcher param = null;

                    param = decodedPattern_ssr_obfsparam.matcher(uri);
                    if (param.find()) {
                        profile.setObfs_param(new String(Base64.decode(param.group(1).replace('/', '_').replace('+', '-').replace("=", ""), Base64.URL_SAFE), StandardCharsets.UTF_8));
                    }

                    param = decodedPattern_ssr_protocolparam.matcher(uri);
                    if (param.find()) {
                        profile.setProtocol_param(new String(Base64.decode(param.group(1).replace('/', '_').replace('+', '-').replace("=", ""), Base64.URL_SAFE), StandardCharsets.UTF_8));
                    }

                    param = decodedPattern_ssr_remarks.matcher(uri);
                    if (param.find()) {
                        profile.setName(new String(Base64.decode(param.group(1).replace('/', '_').replace('+', '-').replace("=", ""), Base64.URL_SAFE), StandardCharsets.UTF_8));
                    } else {
                        profile.setName(ss.group(2).toLowerCase());
                    }

                    param = decodedPattern_ssr_groupparam.matcher(uri);
                    if (param.find()) {
                        profile.setUrl_group(new String(Base64.decode(param.group(1).replace('/', '_').replace('+', '-').replace("=", ""), Base64.URL_SAFE), StandardCharsets.UTF_8));
                    }

                    // add to list
                    list.add(profile);
                }
            }
            return list;
        } catch (Exception e) {
            // Ignore
            VayLog.INSTANCE.e(TAG, "findAll", e);
            ShadowsocksApplication.app.track(e);
            return null;
        }
    }
}

File: app/src/main/java/com/bige0/shadowsocksr/utils/Parser.kt
package com.bige0.shadowsocksr.utils

import com.bige0.shadowsocksr.*
import com.bige0.shadowsocksr.database.*
import java.net.*
import java.util.*

object Parser
{
	const val TAG = "Parser"

	private const val pattern_ss_regex = "(?i)ss://([A-Za-z0-9+-/=_]+)(#(.+))?"
	private const val decodedPattern_ss_regex = "(?i)^((.+?)(-auth)??:(.*)@(.+?):(\\d+?))$"

	private const val pattern_ssr_regex = "(?i)ssr://([A-Za-z0-9+-/=_]+)"
	private const val decodedPattern_ssr_regex = "(?i)^((.+):(\\d+?):(.*):(.+):(.*):([^/]+))"
	private const val decodedPattern_ssr_obfsparam_regex = "(?i)[?&]obfsparam=([A-Za-z0-9+-/=_]*)"
	private const val decodedPattern_ssr_remarks_regex = "(?i)[?&]remarks=([A-Za-z0-9+-/=_]*)"
	private const val decodedPattern_ssr_protocolparam_regex = "(?i)[?&]protoparam=([A-Za-z0-9+-/=_]*)"
	private const val decodedPattern_ssr_groupparam_regex = "(?i)[?&]group=([A-Za-z0-9+-/=_]*)"

	fun findAllSs(data: CharSequence): List<Profile>
	{
		val pattern = Regex(pattern_ss_regex)
		val decodedPattern = Regex(decodedPattern_ss_regex)

		val input: CharSequence = if (data.isNotEmpty()) data else ""
		val list = ArrayList<Profile>()
		try
		{
			pattern.findAll(input)
				.forEach {
					val ss = decodedPattern.matchEntire(Base64.decodeUrlSafe(it.groupValues[1]))
					if (ss != null)
					{
						val profile = Profile()
						val port = ss.groupValues[6].toUShortOrNull()
						if (port != null)
						{
							profile.remotePort = port.toInt()
						}
						else
						{
							return@forEach
						}
						profile.method = ss.groupValues[2].toLowerCase(Locale.ENGLISH)
						if (ss.groups[3] != null)
						{
							profile.protocol = "verify_sha1"
						}
						profile.password = ss.groupValues[4]
						profile.name = ss.groupValues[5]
						profile.host = profile.name
						if (it.groups[2] != null)
						{
							profile.name = URLDecoder.decode(it.groupValues[3], "utf-8")
						}
						list.add(profile)
					}
				}
		}
		catch (e: Exception)
		{
			VayLog.e(TAG, "findAllSs", e)
			ShadowsocksApplication.app.track(e)
		}
		finally
		{
			return list
		}

	}

	fun findAllSsr(data: CharSequence): MutableList<Profile>
	{
		val input = if (data.isNotEmpty()) data else ""
		val list = ArrayList<Profile>()
		try
		{
			Regex(pattern_ssr_regex).findAll(input)
				.forEach {
					val uri = Base64.decodeUrlSafe(it.groupValues[1])
					val ss = Regex(decodedPattern_ssr_regex).find(uri)
					if (ss != null)
					{
						val profile = Profile()
						val port = ss.groupValues[3].toUShortOrNull()
						if (port != null)
						{
							profile.remotePort = port.toInt()
						}
						else
						{
							return@forEach
						}
						profile.host = ss.groupValues[2].toLowerCase(Locale.ENGLISH)
						profile.protocol = ss.groupValues[4].toLowerCase(Locale.ENGLISH)
						profile.method = ss.groupValues[5].toLowerCase(Locale.ENGLISH)
						profile.obfs = ss.groupValues[6].toLowerCase(Locale.ENGLISH)
						profile.password = Base64.decodeUrlSafe(ss.groupValues[7])

						if (profile.obfs == "tls1.2_ticket_fastauth")
						{
							profile.obfs = "tls1.2_ticket_auth"
						}
						var param = Regex(decodedPattern_ssr_obfsparam_regex).find(uri)
						if (param != null)
						{
							profile.obfs_param = Base64.decodeUrlSafe(param.groupValues[1])
						}

						param = Regex(decodedPattern_ssr_protocolparam_regex).find(uri)
						if (param != null)
						{
							profile.protocol_param = Base64.decodeUrlSafe(param.groupValues[1])
						}

						param = Regex(decodedPattern_ssr_remarks_regex).find(uri)
						if (param != null)
						{
							profile.name = Base64.decodeUrlSafe(param.groupValues[1])
						}
						else
						{
							profile.name = profile.host
						}

						param = Regex(decodedPattern_ssr_groupparam_regex).find(uri)
						if (param != null)
						{
							profile.url_group = Base64.decodeUrlSafe(param.groupValues[1])
						}

						list.add(profile)
					}
				}
		}
		catch (e: Exception)
		{
			VayLog.e(TAG, "findAllSsr", e)
			ShadowsocksApplication.app.track(e)
		}
		finally
		{
			return list
		}
	}
}
